waspc/data/Generator/templates/server/src/jobs/PassthroughJobFactory.js

```diff
@@ -0,0 +1,40 @@
+import { sleep } from '../utils.js'
+
+/**
+ * "Immutable-ish" passthrough job wrapper, mainly to be used for testing.
+ */
+class PassthroughJob {
+  constructor(values) {
+    this.perform = () => { }
+    this.delayMs = 0
+    Object.assign(this, values)
+  }
+
+  delay(ms) {
+    return new PassthroughJob({ ...this, delayMs: ms })
+  }
+
+  performAsync(args) {
+    const clonedJob = new PassthroughJob({ ...this })
+    return clonedJob._performAsync(args)
+  }
+
+  _performAsync(args) {
```

Comment:
Hey there! I am 100% rusty on the Promise side lol but I think you understood the motivation in a future comment below. I was hoping to return an actual object (not something we need to resolve) that we can have several other fields on it, perhaps thinks like `jobId` and maybe `jobStatus` or some other stuff from `pg-boss` in future.

The calling syntax (per the Issue) I was going for was:
```js
const runningJob = mySpecialJob.performAsync({ something: "here" })
runningJob.someOtherStuffInFuture
runningJob.result().then(...)
```

I did have `performAsync` and `delay` as class methods to start with, but it required me to use the same name as instance methods for things that can be called on class or chained on instance, so a bit more confusing. Martijn also suggested an immutable instance approach so figured I would give it a go and see.

Also, we need to actually kick off the function in `_performAsync`, so that is why I do not return `fn` itself. This way, when the caller invokes `performAsync` it is actually running **before** they even try to wait for the result (which they may never do). So in my design, `peformAsync` just kicks it off, and `result()` gets the result (possibly blocking for it), but we may have other info/functions on the handle.

As for `result()` vs `result` I don't have enough experience to know the JS convention, but to me it felt better to imply a function. But I defer to ya'll on that.

Reply:
Ok, so this is interesting discussion!

I get now that you want `result` to be promise, in some way. Makes sense to me, as you said, you can return stuff instantly next to it.

But, then you go a step further, and say that you want `result` to be a function that returns promise. That way, people need to call `result()` to even start the execution of the job, which is cool because it becomes "lazy" -> they can choose if they want to run it or not.

But, with the example from before, that seems confusing to me!
With `const runningJob = mySpecialJob.performAsync({ something: "here" })`, I would expect the `runningJob` to be running -> but actually it is not and I have to call `result()` on it to get it running, right? Why did I call `performAsync` if I never wanted to perform it?

What if we instead say that `performAsync(...)` actually performs the job, in async manner of course (which the name says :D!), and then `result` is of type `Promise`, not `fn() : Promise`? Do we lose anything at all, any flexibility? To me that sounds like equally powerful, but more natural way to do this. `result` is not lazy then -> but why would it be, if you didn't want to perform job you shouldn't have called `performAsync`.

Reply:
Well, not quite... they don't need to call `result` to start it, that happens in `performAsync` as intented. They only call `result` if they want a result. The fact that `fn` gets invoked inside `_performAsync` should ensure it actually starts running.

For example, you can see in the output:
```bash
Server: Custom server setup done!
Server: Kicking off Job...
Server: Inside Job bar's callback foo:  { something: 'here' }
Server: Waiting for Job result...
```

We are inside the callback _before_ the line that starts waiting for the result.

Reply:
Aha ok great then -> but then there is no reason for `result` to be a function, right?

Reply:
We chatted and simplified it quite a bit:

```js
  performAsync(args) {
    return {
      result: sleep(this.delayMs).then(() => this.perform(args))
    }
  }
```

There wasn't a true need to clone and call a "private" version really, more overly cautious but not needed. Also the helper function was reduced seeing as how we really didn't need the conditional check. Woot thanks @Martinsos ⛳

Reply:
:golfing_man: :D

Reply:
I need to start thinking more like "and `.then()`," like this lady :D https://youtu.be/CkdyU_eUm1U?t=14

Reply:
Ahahaha ok this was great, I never watched that movie so this was knew to me :D. I still wonder if there was any way to stop the "and then"s!! We will never know I guess.

Reply:
You're going to hate me for commenting this after everything is merged, but I think that awaiting `job.result()` is much better than awaiting `job.result` :sweat_smile:

Although completely valid and functional, it's very unusual to have a regular object field be of type `Promise` (I don't remember ever seeing it). In my experience, libraries always use functions for accessing promises.

This convention most likely comes from the fact that a large part of the JS community views promises as a magic thing returned by asynchronous functions.

Reply:
Hmmm, well if everybody is doing so, it is not a bad argument, but still, it sounds completely redundant, right? To me, also confusing. And I would bet I somewhere dealt with an object that was Promise.
Is there a chance we can find a valid argument for using `result()`, besides everybody else doing it? Btw do you have any examples of smb else doing it?

Maybe you are referring to async functions returning promises?
That of course makes sense, they are async :D, but in a sense we are doing that, we are just being a bit smarter by returning object that had that promise inside it, instead of it all being a promise.

Reply:
Unfortunately, I can't think of an example that uses either approach.

Let's not add a redundant function for now and see if someone will complain. In the end, you are right, there's no reason for it to exist.

Reply:
Don't worry @sodic I will never hate you for commenting after a merge to agree with me :D jk We can leave and see how it feels. As I integrate pg-boss, we may need to tweak this API a bit anyways, but we can keep it in mind. Thanks guys 👍🏻

Reply:
> I will never hate you for commenting after a merge to agree with me

Haha, glad I could help!

BTW, Martin and I discussed the API 10 minutes ago and came up with several more things to keep in mind:
1. Perhaps it makes sense to return the result directly (without keeping it inside the object). Still, I understand the motivation behind the current approach (job metadata and all that).
I am not an expert in the cron/scheduled job area, so I can't precisely estimate the likeliness of encountering specific use cases. @shayneczyzewski, you can be the judge here (and we can tweak it as we go).
2. We were also talking about the best approaches to implement a more advanced API (e.g., periodic jobs) and concluded that we'll most likely want to copy other people :smile: 
More specifically, accessing and controlling jobs from JS is not at all unique to Wasp. There are probably more than a handful of libraries that do precisely that. So it would be wise to check out what some of those tried and tested APIs look like.

Reply:
Yeah, good points to keep in mind for sure! The API is already needing some tweaks for pg-boss integration, as it natively just gives you back a promise that resolves to a job id when you submit a job. If we just return that, we will need to provide some other helpers to look up status, etc. which may be fine. Result could also then be our own helper though to wait until job completion (via their API, hiding the details) but we may still want to give them a result and a job id, and other stuff perhaps? I may pull you both in tomorrow to brainstorm a bit after I have a bit more time to play with it today if that sounds ok.

Reply:
And finally we concluded that we don't know yet, so whatever works we will probably change it anyway as @shayneczyzewski does more work on it :D! So on my side, let's go with whatever makes most sense right now and we will adapt to use cases / pg-boss.

Reply:
Yup @shayneczyzewski soudns perfect!